{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "81448086",
   "metadata": {},
   "source": [
    "# Travail Pratique 6 - Proteomics\n",
    "These exercises will help you to familiarise yourself with different aspects of the analysis of peptide sequences generated by mass spectrometry. You will use a dataset taken from a paper looking at SARS-COV2 proteome interactions (https://doi.org/10.1038/s41586-020-2286-9). Once again, the important thing is to understand the usefulness of the commands you make, and to be able to interpret the type of result obtained in its biological context. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b5cdb40",
   "metadata": {},
   "source": [
    "### Question 1 - Using RforProteomics - File summary of the MS Experiment\n",
    "1. Write down the commands that enabled you to access the data from the TP (symbolic link, cope) and the subsequent working directory indicated in the R script.  \n",
    "    **<font color = green>creat a soft link: [en64851@capemay-sens \\~]# ln -s /cours/a23_bcm3553_bcm6010-a/cours/TP6 \\~/bcm6010/tp6/  \n",
    "Creat a cope for the R script: [en64851@capemay-sens tp6]# cp ./script/TP6_Processing.Rmd \\~/bcm6010/tp6/homework/  \n",
    "In R script: setwd(“\\~/bcm6010/tp6/homework”)</font>**  \n",
    "    .  \n",
    "2. In the first working section (r peptides), what is the full path of the directory from which you take the MS metadata files for MM8 & MM14?  \n",
    "    **<font color = green>filepath \\<- system file(\"microtofq\", package = \"masdata\")  \n",
    "In the Enviroment option in Rstudio, we can check the value of \"filepath\", which is:“/apps_rh8/Local_rh8/R/R-4.3.1/lib64/library/msdata/microtofq”</font>**    \n",
    "    .  \n",
    "3. Looking at the information available for MM8 & MM14, you can see that one of them is just an example file with no peptide. Which one? How do you see it?  \n",
    "    **<font color = green> When performing runInfo(mz14) command, it can be found that althought the \"scanCount\" is 112, both \"lowMz\" and \"highMz\" are 0. So MM14 is an example file with no peptides.</font>**  \n",
    "    .  \n",
    "4. Using the \"runInfo\" command on the non-empty file. What does the second and third values returned mean?  \n",
    "    **<font color = green> When performing runInfo(mz8), it can be found that the \"scanCount\" is 198 (scan number), the \"lowMz\" is 95.51765 and the \"highMz\" IS 1005.043 (represent the lowest and highest mass-to-charge ratio (m/z) value in MS data, respectively). The \"dStartTime\" is 0.486, while the \"dEndTime\" is 66.7818 (start and end time of the scan). \"msLevels\" is 1. </font>**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a60143c1",
   "metadata": {},
   "source": [
    "### Question 2 - Peptide Identification File  \n",
    "1. You are provided with identification files (ID files) for two different MS experiments.  Do they contain exactly the same type of information?  \n",
    "    **<font color = green>No, they do not contain the same type of information.  \n",
    "    when run codes: names(psms(mzidTandem)) and names(psms(mzidMascot)), we can check the types of information (metadata) availabe. It can be found that mzidTandem has 20 types, while mzidMascot has 23.</font>**  \n",
    "    .  \n",
    "2. In the Tandem file, you can find the peptide \"KDLYGNVVLSGGTTMYEGIGER\". What is its experimental mass here? How do you find this information?  \n",
    "    **<font color = green> by running the codes: head(psms(mzidTandem))[,1:13], we can find that the \"experimentalMassToCharge\" for peptide \"KDLYGNVVLSGGTTMYEGIGER\" is 792.3792.</font>**  \n",
    "    .  \n",
    "    \n",
    "3. In the Mascot file, the first peptide to appear is \"DAGTISGLNVLR\", and the first result (covering positions 161-172 of the associated protein) seems to offer the best association. What protein & organism oes this ionized peptide come from?  \n",
    "    **<font color = green> if we use \"heas(psms(mzidMascot))\" command, we could find that the 1st peptid is comming from heat shock cognate 70kDa protein (HSP70) from Echinococcus granulosus.</font>**  \n",
    "    .  \n",
    "    \n",
    "4. Which peptidase was used to prepare the samples for htese two experiments? How does it work?  \n",
    "    **<font color = green> Trypsin: trypsin is a peptidase that can specifically cleave the C-terminal of 2 basic amino acids. The negative charge associated with aspartate (aspartic acid) in trypsin's active site selects for the positive charge of the two basic amino acids lysine and arginine.</font>**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06581959",
   "metadata": {},
   "source": [
    "### Question 3 - Raw MS signal files and their peptide spectral match to proteins  \n",
    "1. looking at the summary of the raw file, how many ionic peaks are associated with spectrum# 2 in the file?  \n",
    "    **<font color = green> there are totally 2921 peaks associated with spectrum# 2 in the file. </font>**  \n",
    "    .  \n",
    "2. Using the graphical representation of peptide spectra, which spectrum has the highest signal intensity, and which has the lowest m/z ratio?  \n",
    "    **<font color = green> in the plots, the y-axis represents the signal intensity, while the x-axis represnets the m/z ratio.  \n",
    "    From the plots, we can find that spectrum# 1 has the higest signal intensity (~1.5e+07), and the spectrum# 5 has the lowest m/z ratio (~750).</font>**    \n",
    "    .  \n",
    "3. what are iTRAQ reporter here for?  \n",
    "    **<font color = green> iTRAQ = isobaric tags for relative and absolute quantitation. Isobaric tags for relative and absolute quantitation (iTRAQ) allow for amine-specific isobaric tagging and thus are suitable for simultaneous analysis of proteins regardless of the presence of cysteine residues (Vélez-Bermúdez, et al. 2016) </font>**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f00be5e",
   "metadata": {},
   "source": [
    "### Question 4 - Reconstruction of peptide data and graphical visualisation  \n",
    "1. How many samples (not proteins) were analysed?  \n",
    "    **<font color = green> By running the code \"processingData(qnt)\", we can see a subset with 1528 rows and 6 colunms was created during the analysis. 1528 rows represent the values of the measurements which indicate the sample number.</font>**  \n",
    "    .  \n",
    "    \n",
    "2. The signals com from known protein peptides (eg. PYGM_Rabit), with known concentrations. Looking at the first matplot graph, which protein had been diluted as follow across the data: 1/10 -- 1/4 -- 1/2 -- 1 -- 1/2 -- 1/10  \n",
    "    **<font color = green> sample 4 - ALBU-BOVIN </font>**  \n",
    "    .  \n",
    "    \n",
    "3. Now looking at the heatmap, why do we no longer seem to be interested in \"TRYP_PIG\" and why do we add \"ECA\" (Enterobacterial Common Antigen).  \n",
    "    **<font color = green> We started to focus on the specific samples and its protein expressed. It can be seen that each column represents one reporter (126, 127, 128, 129, 130, 131). specific proteins (UniProt codes) are selected, and the top 10 features (peptides) associated with each protein are extracted.</font>**  \n",
    "    .  \n",
    "    \n",
    "4. which reporter has the highest concentration of Bovine Albumin?  \n",
    "    **<font color = green> From the heatmap, reporter 129 has the highest concentration of bovine albumin. </font>**   \n",
    "    .  \n",
    "    \n",
    "5. Why can we only confirm the presence in 131, despite the fact that a CYT signal is found in 4 reporters according to the heatmap?  \n",
    "    **<font color = green> Firstly, there are only 2 measurements for CYR in each reporter. Secondly, from last plot we could observe that the variation is huge for the other reporter.</font>**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6055172a",
   "metadata": {},
   "source": [
    "### Question 5 - Data Processing for GO Enrichment  \n",
    "We are now going to start looking at the data from the SARS-CoV2 paper mentioned above. When an article provides an interesting approach to analysis, it may be appropriate to apply the methodology to our own data. Since the scripts are not directly provided by the article, and therefore have to be replicated manually, it is a good idea to use the data from the article to validate our scripts by replicating the published results. The questions follow the order of the code.\n",
    "1. Basic question about file formats: Without looking at the contents of \"Gordon_PPI.tsv\", how do we know that \"sep = \"\\t\"\" is required for reading ?\n",
    "    **<font color =green> According to the file type shown in the file name, “Gordon_PPI.tsv” is a tsv file (test-based) whose contents are separated by tab. Thus, it is necessary to use the command “sep = “\\t” to read through the file.</font>**  \n",
    "    .  \n",
    "2. We extract the GO terms that we are going to use in \"go_symbols\". We are using \"C5.GO\", but what other type of terms could we have used given that we are interested in molecular pathways associated with normal cell functions (google the name of the file).\n",
    "    **<font color = green> C5 GO gene sets consist of genes annotated by the same GO terms (Martin, et al.). Given that we are interested in molecular pathways associated with normal cell function, we could also use C2 Curated Gene sets, which are from online pathway databases and publications. </font>**  \n",
    "    .  \n",
    "3. We extract 3 types of data for our GO terms (exactSource, term, collection), but how many were available in total in the annotation file?  \n",
    "    **<font color = green> there are totally 10 tyoes of data available in the annotation file.</font>**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b627488a",
   "metadata": {},
   "source": [
    "### Question 6 - Understanding our data \n",
    "1. In total, the proteomics experiment collected signals from thousands of peptides, which could be reconstructed into 304 proteins (2-40 per SARS-CoV2 protein). Just to make sure you understand the experiment. How many 'Prey' proteins are there? How many Bait proteins? Name three of these Baits. We're at the 1st line of processing_GO code.  \n",
    "    **<font color = green>The Prey proteins are the proteins that interact with the SARS-CoV2 proteins (Baits) in the experiment.There are 304 prey proteins and 25 bait proteins.</font>**  \n",
    "    .  \n",
    "2. Explain the purpose of the \"unique\" function, why is it used here?  \n",
    "    **<font color = green>The unique() function could avoid repetitive count of the same protein in the dataset.</font>**  \n",
    "    .  \n",
    "3. In our \"GO_OverRep_Analysis\" function, we process the results of the clusterProfiler tool in one go. Explain briefly what the min/maxGSsize thresholds are used for, and then explain in what format the significance values are given by the function.  \n",
    "    **<font color = green> According to the R script, the min GS size is 10, the max GS size is 500. These two parameters are used for filter the size of the gene sets. In this Rscript, only those  gene sets contains more than 10 but smaller than 500 genes will be shown.   \n",
    "    The significance values are setted by \"pvalueCutoff\" as 0.05 and will be adjusted by Benjamini-Hochberg test. The significance values are given as negative logrithms (-log10).</font>**  \n",
    "    .  \n",
    "4. usng a line of code, we remove all the GO terms where the score for our data is equal to or less than 1 from \"enriched_GO_master_table\". How many GO rows are removed?  \n",
    "    **<font color = green> before the removment, there are 7398 entries; after the removment, there are only 1361 entries. Totally 6037 GO rows are removed.</font>**  \n",
    "    .  \n",
    "5. How many GO terms do not pass the p-value filter?  \n",
    "    **<font color = green> 329 GO terms do not pass the p-value filter.</font>**  \n",
    "    .  \n",
    "6. In the \"reduce_GOredundancy\" function, we call up a database containing human GO genes, which we have already used. What is the name of this database?  \n",
    "    **<font color = green> C5.GO </font>**  \n",
    "    .  \n",
    "7. Here are some GO terms. Determine whether they are biological process(BP), molecular function(MF) or cellular component(CC) GOs.  \n",
    "    **<font color = green> Glycoprotein metabolic processing - BP  \n",
    "    Golgi apparatus subcompartment - CC  \n",
    "    DNA replication initation - BP  \n",
    "    Endopeptidase activity - MF</font>**  \n",
    "    .  \n",
    "8. Once the redundancy has been filtered via \"apply_GO_nonRedundancy\", use the View function to find out how many GO terms are associated with NSP7, and name one of them.  \n",
    "    **<font color = green> Totally 220 GO terms are associated with NSP7, for example: GOBP_Polysaccharide_Biosynthetic_process </font>**  \n",
    "    .  \n",
    "9. Our GO term typing function (sep_GOtype) allows us to analyse our data more simply depending on what we are looking for (e.g. biological processes appear to be more revealing). Following its application, how many CC-type GOs do we have for ORF6? Name one.\n",
    "    **<font color = green> There is only one CC-type GO term for ORF6, it is GOCC_Nuclear_Pore. </font>**  \n",
    "    .  \n",
    "10. A little later, we apply the \"base::order\" function to sort our data in the matrix according to their significance. If you look at the result of this transformation (you can use View()), what is the most significant GO term, to which \"Bait\" does it belong, how many \"Prey\" proteins are involved in the enrichment signal, name one of these \"Prey\", and then give the significance value in log10 of the adjusted p-value.  \n",
    "    **<font color = green> According to the Rscript, “enriched_GOs_noRe_DF” is sorted in descending order by the -log10 p-value by the function “base::order”. In this case, the most significant GO term should be “GOCC_Endoplasmic_Reticulum_Lumen”. The adjusted log10_p is 13.90. It is belonged to ORF8. There are 16 prey proteins involved in the enrichment signal.</font>**  \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "771dd5e0",
   "metadata": {},
   "source": [
    "### Question 7 - Preparing data for the Heatmap\n",
    "1. if we initialize the SARS_genes_count variable at the start of the (empty) section, and add SARS_gene to it following the three conditions in our \"loop\", what is the purpose of this condition at the start of the loop: while (length(SARS_genes_count) <= 25)? This question is purely a test of your understanding of R code. The answer you're looking for is simple and doesn't need any biological interpretation.   \n",
    "    **<font color = green>While (length(SARS_genes_count)<=25) ——This while loop will continue until the number of elements in “SARS_genes_count” reaches 25.</font>**  \n",
    "    .  \n",
    "    \n",
    "2. The GO_enrich_DF matrix is ultimately used to create the Heatmap. Why is the significance value transformed?   \n",
    "    **<font color = green> Transform the p value to the -log10(), so that we can use negative logarithms as a colour metric in heatmaps. Generally, p-value is between 0-1, this transformation can convert the value to 0-∞. The smaller p-value (more significant) has larger -log10 results. In the figure, this can help to highlight the significant level. </font>**  \n",
    "    .  \n",
    "    \n",
    "3. Simply explain what these two lines do in the code:  \n",
    "    rownames_matrix <- GO_enrich_matrix[,1]  \n",
    "    GO_enrich_matrix <- GO_enrich_matrix[,-c(1)]  \n",
    "    **<font color = green> Assign the first column in “GO_enrich_matrix” (which should be “GO_term\") to the new variable “rownames_matrix”. After this step, the first column is deleted from the “GO_enrich_matrix” by using “[,-c(1)]”. </font>**  \n",
    "    .  \n",
    "4. How many colors are in \"colors3\"?  \n",
    "    **<font color = green> According to the code: colors3 <- colorRampPalette(brewer.pal(9, \"Blues\"))(256)  \n",
    "                                                             9 colors from “Blues” palettes were selected.  These colours are extended to 256 colours with “colorRampPalette()” function</font>**  \n",
    "    .  \n",
    "5. Which heatmap parameter allows us to change the width of the heatmap?  \n",
    "    **<font color = green>We could add “cellwidth = value” to adjust the width of the heatmap.</font>**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7de0635",
   "metadata": {},
   "source": [
    "### Question 8 - Check the Results  \n",
    "1. According to the heatmap, does the most prominent of these align with what we identified in question 6 j)?\n",
    "    **<font color = green>In 6(j), we observed the most significant GO term is \"GOCC_Endoplasmic_Reticulum_Lumen\", which is belong to ORF8. Here from the heatmap, we could also observe that ORF8 has the most enriched GO term-endoplasmic reticulum lumen. This is consistent with 6(j).</font>**  \n",
    "    .  \n",
    "2. Going back to the presentation of the article at the start of the TP, if we look at the most enriched GO for the \"Bait\"-“NSP12\" (albeit very pale compared to other GO-SARS links), does the result make sense/is it interesting?   \n",
    "    **<font color = green>According to the article, NSP12 is the primary RNA-dependent RNA polymerase. NSP12 may be involved in some specific and relatively narrow biological processes that are not involved in the broader pathways associated with other SARS proteins. This may result in a lower enrichment of GO terms associated with NSP12 compared to other SARS proteins.</font>**  \n",
    "    .  \n",
    "3. Why is there a relatively large blue signal for 3 GO terms opposite the \"Bait\"-\"NSP9\"? It's not a mistake, just think about how the GOs are chosen.  \n",
    "    **<font color = green>Nsp9 is a non-structure protein, which is though to mediate viral replication, overall virulence, and viral genomic RNA reproduction (Littler, et al. 2020). The enriched GO terms include Mitochondrial transport, nuclear pore and structural constituent of nuclear pore. This also consistent with its predicted function which associated with the process of viral replication and infection within the host cells.Additionally, C5.GO dataset covers many biological processes, molecular functions and cellular components. This allows enrichment analyses to involve multiple aspects of biological</font>**  \n",
    "    .  \n",
    "4. Looking again at \"NSP9\", what clear conclusion can be drawn about the potential function of this SARS-CoV2 protein?  \n",
    "    **<font color = green>It can be found that NSP9 has relatively intensive blue signal for mitochondrial transport, nuclear pore and  structural constituent of nuclear pore. The function of NSP9 might associated with the process of viral replication and infection within the host cells. </font>**  \n",
    "    .  \n",
    "5. Why do some Baits have such pale signals for almost all GOs?  \n",
    "    **<font color = green>One possible reason is that the p-value cutoff or the correction methods we chosed affect the significance of these Baits genes. </font>**  \n",
    "    .  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77b196ad",
   "metadata": {},
   "source": [
    "### Question 9 - Data for Bar Plot\n",
    "1. We store the information from the \"GTEx_Tissu_Abundance.xlsx\" file in the \"proteome_tissue\" matrix. Describe the information contained in your matrix.    \n",
    "    **<font color = green> This matrix contains 32 columns, including the protein ID, gene ID, gene name (totally 15,210 genes), and the gene abundance of each gene in different tissues (29 tissues). </font>**  \n",
    "    .  \n",
    "2. Explain what the transformation we make to our data seems to change.  \n",
    "    **<font color = green> Again, we perform the logarithmic transformation for the Gene expression abundance. This transformation can help to adjust the distribution of the data so that it can fit the Gaussian’s distribution for statistic analysis.</font>**  \n",
    "    .  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04b7d305",
   "metadata": {},
   "source": [
    "### Question 10 - Tissue data Visualisation\n",
    "1. Interpret in a few words what the graph shows us.  \n",
    "    **<font color = green> The bar chart shows the expression level of SRAS-related proteins in different tissues. The X-axis represents different tissues, while Y-axis represents the -long10 p-value which indicates the expression abundance. The higher the bar is, the smaller the p-value it has. From the bar chart, lung has the smallest p-value for the SARS-related gene expression abundance compared to the expression abundance of all the gene in lung, however, the difference is not significant.</font>**  \n",
    "    .  \n",
    "2. Based on a brief review of the literature (with supporting citations), are your results consistent with what is expected of SARS-CoV2?   \n",
    "    **<font color = green> From the bar plot we observed that the most significant SARS-gene expression in lung, which is in line with the expectations since SARS-CoV-2 predominantly attacks the respiratory tract. In addition, a growing number of studies of tissue and autopsy collected from patients have shown that SARS-CoV-2 infection is present in multiple organs due to viral replication (Yu, et al. 2022). This also consistent with our heatmap and bar plot.</font>**  \n",
    "    .  \n",
    "3. Given that a result is normally recognised as \"significant\" if p < 0.05, are your results here significant? Explain in a few words  \n",
    "    **<font color = green> When p = 0.05, -log10(0.05) = 1.301. Thus, the result can be recognized as “significant” of -log10(p-value) > 1.301. However, from the bar chart, we can see that the highest -log10 p-value is around 0.6 which represents the Lung’s protein expression level. In conclusion, my results are not significant.</font>**  \n",
    "    .  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3486e213",
   "metadata": {},
   "source": [
    "#### REFERENCE  \n",
    "1. Vélez-Bermúdez IC, Wen TN, Lan P, Schmidt W. Isobaric Tag for Relative and Absolute Quantitation (iTRAQ)-Based Protein Profiling in Plants. Methods Mol Biol. 2016;1450:213-21. doi: 10.1007/978-1-4939-3759-2_17. PMID: 27424757.  \n",
    "2.  Littler DR, Gully BS, Colson RN, Rossjohn J. Crystal Structure of the SARS-CoV-2 Non-structural Protein 9, Nsp9. iScience. 2020 Jul 24;23(7):101258. doi: 10.1016/j.isci.2020.101258. Epub 2020 Jun 9. PMID: 32592996; PMCID: PMC7282741.  \n",
    "3. Yu S, Li X, Xin Z, Sun L, Shi J. Proteomic insights into SARS-CoV-2 infection mechanisms, diagnosis, therapies and prognostic monitoring methods. Front Immunol. 2022 Sep 20;13:923387. doi: 10.3389/fimmu.2022.923387. PMID: 36203586; PMCID: PMC9530739."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
